SSA/WellTypedFramework.lean

```diff
@@ -1,57 +1,60 @@
 
-def toType {BaseType : Type} [DecidableEq BaseType] [BaseTypeClass BaseType] : UserType BaseType → Type
+def toType [BaseTypeClass BaseType] : UserType BaseType → Type
   | .base t => BaseTypeClass.toType t
   | .pair k₁ k₂ => (toType k₁) × (toType k₂)
   | .triple k₁ k₂ k₃ => toType k₁ × toType k₂ × toType k₃
   | .unit => Unit
 
-def mkPair {BaseType : Type} [DecidableEq BaseType] [BaseTypeClass BaseType] {k₁ k₂ : UserType BaseType}: toType k₁ → toType k₂ → toType (.pair k₁ k₂)
+def mkPair {BaseType : Type} [BaseTypeClass BaseType] {k₁ k₂ : UserType BaseType}: toType k₁ → toType k₂ → toType (.pair k₁ k₂)
   | k₁, k₂ => (k₁, k₂)
 
-def mkTriple {BaseType : Type} [DecidableEq BaseType] [BaseTypeClass BaseType] {k₁ k₂ k₃ : UserType BaseType}: toType k₁ → toType k₂ → toType k₃ → toType (.triple k₁ k₂ k₃)
+def mkTriple {BaseType : Type} [BaseTypeClass BaseType] {k₁ k₂ k₃ : UserType BaseType}: toType k₁ → toType k₂ → toType k₃ → toType (.triple k₁ k₂ k₃)
   | k₁, k₂, k₃ => (k₁, k₂, k₃)
 
-def fstPair {BaseType : Type} [DecidableEq BaseType] [BaseTypeClass BaseType] {k₁ k₂ : UserType BaseType} : toType (.pair k₁ k₂) → toType k₁
+def fstPair {BaseType : Type} [BaseTypeClass BaseType] {k₁ k₂ : UserType BaseType} : toType (.pair k₁ k₂) → toType k₁
   | (k₁, _) => k₁
 
-def sndPair {BaseType : Type} [DecidableEq BaseType] [BaseTypeClass BaseType] {k₁ k₂ : UserType BaseType} : toType (.pair k₁ k₂) → toType k₂
+def sndPair {BaseType : Type} [BaseTypeClass BaseType] {k₁ k₂ : UserType BaseType} : toType (.pair k₁ k₂) → toType k₂
   | (_, k₂) => k₂
 
-def fstTriple {BaseType : Type} [DecidableEq BaseType] [BaseTypeClass BaseType] {k₁ k₂ k₃ : UserType BaseType} : toType (.triple k₁ k₂ k₃) → toType k₁
+def fstTriple {BaseType : Type} [BaseTypeClass BaseType] {k₁ k₂ k₃ : UserType BaseType} : toType (.triple k₁ k₂ k₃) → toType k₁
   | (k₁, _, _) => k₁
 
-def sndTriple {BaseType : Type} [DecidableEq BaseType] [BaseTypeClass BaseType] {k₁ k₂ k₃ : UserType BaseType} : toType (.triple k₁ k₂ k₃) → toType k₂
+def sndTriple {BaseType : Type} [BaseTypeClass BaseType] {k₁ k₂ k₃ : UserType BaseType} : toType (.triple k₁ k₂ k₃) → toType k₂
   | (_, k₂, _) => k₂
 
-def trdTriple {BaseType : Type} [DecidableEq BaseType] [BaseTypeClass BaseType] {k₁ k₂ k₃ : UserType BaseType} : toType (.triple k₁ k₂ k₃) → toType k₃
+def trdTriple {BaseType : Type} [BaseTypeClass BaseType] {k₁ k₂ k₃ : UserType BaseType} : toType (.triple k₁ k₂ k₃) → toType k₃
   | (_, _, k₃) => k₃
 
 end UserType
 
 
-structure UserData {BaseType : Type} [DecidableEq BaseType] [BaseTypeClass BaseType] where
+structure UserData {BaseType : Type} [BaseTypeClass BaseType] where
   type : UserType BaseType
   value : type.toType
 
 
-inductive  TypeData (BaseType : Type) [DecidableEq BaseType] [BaseTypeClass BaseType] : Type
+-- @goens: I suggest removing [DecidableEq ...] [BaseTypeClass ...]
+-- We do not keep typeclass constraints on definitions, only on use sites.
+-- See the Haskell rationale: 
```

Comment:
Interesting! My rationale was that while `DecidableEq` is technically unnecessary for `BaseTypeClass`, it's morally what we want, same for `TypeData`, we morally want this only for user types, that's why I put them here. I think you forgot to put the link to the Haskell rationale? I'd be curious to understand why.